src/topics/resolver/telegram-resolver.ts

```diff
@@ -0,0 +1,160 @@
+import { BigNumberish } from "ethers";
+import { Api, TelegramClient } from "telegram";
+import { LogLevel } from "telegram/extensions/Logger";
+import { StringSession } from "telegram/sessions";
+import { IResolver } from "./resolver";
+import {
+  resolveAccount,
+  withConcurrency,
+  handleResolvingErrors,
+} from "./utils";
+import { FetchedData } from "topics/group";
+
+export class TelegramResolver implements IResolver {
+  private _apiHash: string;
+  private _apiId: string;
+  private _botSession: string | undefined;
+  private _botToken: string;
+  private _client: TelegramClient;
+
+  constructor(
+    apiHash = process.env.TELEGRAM_API_HASH || "",
+    apiId = process.env.TELEGRAM_API_ID || "",
+    botToken = process.env.TELEGRAM_BOT_TOKEN || "",
+    botSession = process.env.TELEGRAM_BOT_SESSION
+  ) {
+    this._apiHash = apiHash;
+    this._apiId = apiId;
+    this._botSession = botSession;
+    this._botToken = botToken;
+  }
+
+  public resolve = async (
+    accounts: FetchedData
+  ): Promise<[FetchedData, FetchedData]> => {
+    const alreadyUpdatedAccounts: FetchedData = {};
+    const alreadyResolvedAccounts: FetchedData = {};
+
+    // extract usernames already resolved
+    const unresolvedAccounts = Object.entries(accounts).filter(
+      ([account, value]) => {
+        if (account.split(":").length === 3) {
+          const id = account.split(":")[2];
+          alreadyResolvedAccounts[resolveAccount("1003", id)] = value;
+          alreadyUpdatedAccounts[account] = value;
+        }
+        return account.split(":").length !== 3;
+      }
+    );
+
+    await this._initTelegramClient();
+    await this._client.connect();
+
+    const resolvedAccountsArrays = await withConcurrency(
+      unresolvedAccounts,
+      this.resolveAccounts,
+      {
+        concurrency: 10,
+        batchSize: 1,
+      }
+    );
+
+    await this._client.disconnect();
+
+    // merge already resolved accounts with the new ones
+    const resolvedAccountsRaw = {
+      ...resolvedAccountsArrays[0],
+      ...alreadyUpdatedAccounts,
+    };
+    const resolvedAccounts = {
+      ...resolvedAccountsArrays[1],
+      ...alreadyResolvedAccounts,
+    };
+
+    return [resolvedAccountsRaw, resolvedAccounts];
+  };
+
+  private _initTelegramClient = async () => {
+    this._client = new TelegramClient(
+      new StringSession(this._botSession), +this._apiId, this._apiHash, {}
+    );
+    this._client.setLogLevel(LogLevel.ERROR);
+    await this._signInIfNeeded();
+  };
+
+  private _signInIfNeeded = async () => {
+    if (this._botSession) {
+      console.log("Session found, no need to authorize bot.")
```

Comment:
I think this log is not needed, simply use bot token if the session is not referenced seems good to me

Reply:
It's a bit verbose indeed, I'll remove it. Thanks!